keps/420-partial-admission/README.md

```diff
@@ -0,0 +1,163 @@
+
+### User Stories
+
+Kueue issue [420](https://github.com/kubernetes-sigs/kueue/issues/420) provides details on the initial feature details and its applicability for `batch/Job`.
+
+## Design Details
+
+### Workload API
+
+```go
+type PodSet struct {
+    // .......
+
+    // count is the number of pods for the spec.
+    // +kubebuilder:validation:Minimum=1
+    Count int32 `json:"count"`
+
+    // minimumCount is the minimum number of pods for the spec acceptable
+    // in case of partial admission.
+    //
+    // If not provided, partial admission for the current PodSet is not
+    // enabled.
+    // +optional
+    MinimumCount *int32 `json:"minimumCount,omitempty"`
+}
+
+```
+
+### Scheduler / Flavorassignment
+
+In case the workload proposed for the current scheduling cycle, does not fit, with or without preemption, in the current available quota and any of its PodSets allow partial admission, try to find to find a lower counts combination that fits the available quota.
+
+The search should be optimized (binary search) and preserve the proportion of pods lost across the variable count PodSets.
+
+Starting with the PodSets:
+
+```go
+[]podSet { {Count: 1} , {Count: 4, MinCount: 2}, {Count: 20, MinCount: 10}}
+```
+
+And only being able to admit 19 pods, it should end up with 
+
+
+```go
+[]podSet { {Count: 1} , {Count: 3}, {Count: 15}}
+```
+
+With both pods that accept partial admission losing 50% of their range.
+
+
+The accepted number of pods in each PodSet are recorded in `workload.Status.Admission.PodSetAssignments[*].ResourceUsage.Count`
+
+### Jobframework
+
+```diff
+type GenericJob interface {
+    // ...
+
+	
+-    // RunWithNodeAffinity will inject the node affinity extracting from workload to job and unsuspend the job.
++    // RunWithPodSetsInfo will inject the node affinity and podsSet counts extracted from workload to the job and unsuspend the job.
+-    RunWithNodeAffinity(nodeSelectors []PodSetNodeSelector)
++    RunWithPodSetsInfo(nodeSelectors []PodSetNodeSelector, podSetCounts []int32)
+-    // RestoreNodeAffinity will restore the original node affinity of job.
++    // RestorePodSetsInfo will restore the original node affinity of job.
+-    RestoreNodeAffinity(nodeSelectors []PodSetNodeSelector)
++    RestorePodSetsInfo(nodeSelectors []PodSetNodeSelector, podSetCounts []int32)
+
+    // ...
+}
+
+```
+
+### batch/Job controller
+
+Besides adapting `RunWithPodSetsInfo` and `RestorePodSetsInfo` it should also:
+
+- rework `PodSets()` to populate `MinimumCount` if the job is marked to support partial admission.
+  * jobs supporting partial admission should have a dedicated annotation. eg. `kueue.x-k8s.io/job-min-parallelism`, indicating the minimum `parallelism` acceptable by the job in case of partial admission.
+  * jobs which need the `completions` count kept in sync with `parallelism` should indicate this in a second annotation `kueue.x-k8s.io/sync-completions`
```

Comment:
```suggestion
  * jobs which need the `completions` count kept in sync with `parallelism` should indicate this in a second annotation `kueue.x-k8s.io/job-completions-equal-parallelism`
```

@ahg-g wdyt?

Reply:
Perhaps the integration controller defaults `kueue.x-k8s.io/job-completions-equal-parallelism` to true if the job was created with `parallelism=completions`.

Reply:
I'm not sure we can imply this. You could have an embarrassingly parallel job where you ideally want maximum parallelism, but are ok with less.
I think it's better for the user to be explicit about the semantics they want, given that parallelism and completions are separate fields to begin with

Reply:
but yeah, it could be surprising

Reply:
it's a bit counter-intuitive to me,  if `parallelism=completions` is the most common use-case then , in my opinion, we should have an opt-out annotation something like `kueue.x-k8s.io/dont-sync-completions`...

Let's keep the opt-in `kueue.x-k8s.io/job-completions-equal-parallelism` for now.